transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,598 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Background task dedicated to manage the QUIC state machine.
+//!
+//! Considering that all QUIC communications happen over a single UDP socket, one needs to
+//! maintain a unique synchronization point that holds the state of all the active connections.
+//!
+//! The endpoint represents this synchronization point. It is maintained in a background task
+//! whose role is to interface with the UDP socket. Communication between the background task and
+//! the rest of the code only happens through channels. See the documentation of the
+//! [`EndpointDriver`] for a thorough description.
+
+use crate::{
+    connection::Connection,
+    tls,
+    transport::{self, Provider},
+};
+
+use futures::{
+    channel::{mpsc, oneshot},
+    prelude::*,
+    ready,
+};
+use std::{
+    collections::HashMap,
+    io,
+    net::{Ipv4Addr, Ipv6Addr, SocketAddr},
+    ops::ControlFlow,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+use x509_parser::nom::AsBytes;
+
+/// Represents the configuration for the QUIC endpoint.
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// The client configuration to pass to `quinn_proto`.
+    client_config: quinn_proto::ClientConfig,
+    /// The server configuration to pass to `quinn_proto`.
+    server_config: Arc<quinn_proto::ServerConfig>,
+    /// The endpoint configuration to pass to `quinn_proto`.
+    endpoint_config: Arc<quinn_proto::EndpointConfig>,
+}
+
+impl Config {
+    /// Creates a new configuration object with default values.
+    pub fn new(keypair: &libp2p_core::identity::Keypair) -> Result<Self, tls::ConfigError> {
+        let mut transport = quinn_proto::TransportConfig::default();
```

Comment:
Is there any config knob for stream / connection flow control (initial) window sizes? What are the defaults that quinn uses? Does quinn do auto-scaling of receive windows, where maximum values might need to be configured?

Reply:
These are default values https://github.com/quinn-rs/quinn/blob/main/quinn-proto/src/config.rs#L257.

I don't know anything about auto-scaling, it would be nice to dig into it.

Reply:
Yeah there is a CUBIC(default) and New Reno congestion controllers.

Reply:
Thank you for the link, @kpp!

Looking at these values, we should probably modify:
* `allow_spin` to `false`: The spin bit exposes the end-to-end RTT to passive on-path observers. When run over a VPN or other tunneling solutions, this exposes _more_ information than a TCP connection.
* the stream receive window is set to 1.25 MB. This seems a lot as a starting value, but very little if this is the maximum value the window can ever grow to. Depends on the auto-scaling logic, if there is any.
* the connection receive window apparently is set to infinity. This is not a very good idea, since it means that the memory you're committing to the QUIC connection is the product of your stream receive window and the number of streams. With default values, that's 125 MB, making it very easy to perform a DoS against the QUIC stack.

Reply:
> `allow_spin` to false

Included in 2264ab83377758d0644e668caba1ae9048cc384c.

> the stream receive window is set to 1.25 MB. This seems a lot as a starting value, but very little if this is the maximum value the window can ever grow to. Depends on the auto-scaling logic, if there is any.

From what I understand in the quinn-proto code it scales to that the actual stream receive window is `stream_receive_window` + bytes read thus far on the stream [^1]

> the connection receive window apparently is set to infinity. This is not a very good idea, since it means that the memory you're committing to the QUIC connection is the product of your stream receive window and the number of streams. With default values, that's 125 MB, making it very easy to perform a DoS against the QUIC stack

What default is go-quic using? Does it scale or is it set to a fixed value?

[^1]: https://github.com/quinn-rs/quinn/blob/cd2cf756acc9e7774d2c40992e7a355d0fb51b74/quinn-proto/src/connection/streams/recv.rs#L99

Reply:
quic-go, not go-quic ;)

The starting value for the connection window is 768 kB, and we scale it up if needed, up to 15 MB. ([code](https://github.com/libp2p/go-libp2p/blob/63b8803005ecca56db249a636453e19158499957/p2p/transport/quic/transport.go#L41-L52))

Reply:
Discussed out of band with @marten-seemann: 4cad0803d270bf3e93104ef723e5b94016c475f9 sets the receive window for a connection to 15MB and for streams to 10MB.